### PR TITLE
Documentation of configuration

### PR DESCRIPTION
**Pull request by beam2d**
This PR adds a documentation for `chainer.config` and modifies some existing docs related to any configuration entries (e.g. debug mode document, tutorial).

### CODE REVIEW

File: docs/source/reference/core/configuration.rst [hunk 1]
```diff
@@ -0,0 +1,102 @@
...
+..  currentmodule:: chainer
+
+Chainer provides some global settings that affect the behavior of some functionalities.
+Such settings can be configured using the *unified configuration system*.
+The system provides a transparent way to manage the configuration for each process and for each thread.
+
+The configuration is managed by two global objects: :data:`chainer.global_config` and :data:`chainer.config`.
+The :data:`global_config` object maintains the configuration shared in the Python process.
+This is an instance of the :class:`~chainer.configuration.GlobalConfig` class.
+It can be used just as a plain object, and users can freely set any attributes to it.
```
Comment (bkvogel) [inline docs/source/reference/core/configuration.rst]: to -> on:
set any attributes on it.

File: docs/source/reference/core/configuration.rst [hunk 2]
```diff
@@ -0,0 +1,102 @@
...
+Note that the default values are set to the global config.
+
+``chainer.config.debug``
+   Debug mode flag.
+   If it is ``True``, Chainer runs in the debug mode.
+   See :ref:`debug` for more information of the debug mode.
+   The default value is given by ``CHAINER_DEBUG`` environment variable (set to 0 or 1) if available, otherwise uses ``False``.
+``chainer.config.enable_backprop``
+   Flag to enable backpropagation support.
+   If it is ``True``, the default behavior of :class:`Function` application to :class:`Variable` is non-volatile if all inputs has ``AUTO`` volatile flag.
```
Comment (bkvogel) [inline docs/source/reference/core/configuration.rst]: if all inputs have

File: docs/source/reference/core/configuration.rst [hunk 3]
```diff
@@ -0,0 +1,102 @@
...
+It can be used just as a plain object, and users can freely set any attributes to it.
+The :data:`config` object, on the other hand, maintains the configuration for the current thread.
+It behaves like a thread-local object, and any attribute modifications are only visible to the current thread.
+If no value is set to this thread-local configuration for a given key, the global configuration is transparently referred.
+Thanks to this transparent lookup, users can always use the ``config`` object to read any configuration so that the thread-local configuration is used if available and otherwise the default global setting is used.
+The ``config`` object is an instance of the :class:`~chainer.configuration.LocalConfig` class.
+
+The following entries of the configuration are currently provided by Chainer.
+Some entries support environment variables to set the default values.
+Note that the default values are set to the global config.
```
Comment (bkvogel) [inline docs/source/reference/core/configuration.rst]: are set in the global config.

File: docs/source/reference/core/configuration.rst [hunk 4]
```diff
@@ -0,0 +1,102 @@
...
+
+   If you set an attribute to the local configuration, the value is only visible to the current thread.
+
+   .. doctest::
+
+      >>> chainer.config.user_my_setting = 123
+
+   We often want to temporarily modify the configuration for the current thread.
+   It can be done by using :func:`using_config`.
+   For example, if you want to enable the debug mode only to a fragment of codes, write as follows.
```
Comment (bkvogel) [inline docs/source/reference/core/configuration.rst]: For example, if you only want to enable debug mode in a fragment of code, write as follows:

File: docs/source/reference/core/configuration.rst [hunk 5]
```diff
@@ -0,0 +1,102 @@
...
+
+      >>> with chainer.using_config('debug', True):
+      ...     ...  # code running in the debug mode
+
+   We often want to switch to the test mode for an evaluation.
+   This is also done by the same way.
```
Comment (bkvogel) [inline docs/source/reference/core/configuration.rst]: This is also done in the same way.

File: docs/source/tutorial/basic.rst [hunk 6]
```diff
@@ -559,6 +559,7 @@ These extensions perform the following tasks:
 
 :class:`~training.extensions.Evaluator`
    Evaluates the current model on the test dataset at the end of every epoch.
+   It automatically switches to the test mode (see :ref:`configuration` for details), and so we do not have to take care on functions that behave differently in training/test modes (e.g. :func:`~chainer.functions.dropout`, :class:`~chainer.links.BatchNormalization`).
```
Comment (bkvogel) [inline docs/source/tutorial/basic.rst]: "take care" seems a bit unclear. Do you mean:

and so we do not have to take any special action for functions that behave differently

File: docs/source/reference/core/configuration.rst [hunk 7]
```diff
@@ -0,0 +1,102 @@
...
+
+      >>> with chainer.using_config('train', False):
+      ...     ...  # code running in the test mode
+
```
Comment (delta2323) [inline docs/source/reference/core/configuration.rst]: The instruction on how to write functions or links that behave differently in train and test modes would be beneficial for developers. Something like this:

```
If you want to write a Chainer function or link that behave differently in train and other modes. You can switch their behavior with the value of the configuration (`chainer.config.train` by default).

class MyLink(chainer.Link):

    def __call__(self, x):
        if chainer.config.train:
            #  Implement training mode.
        else:
            #  Implement evaluation mode.
```

**Comment by delta2323** [comment]
It would be helpful for users the previous way of setting configuration (e.g. `chainer.set_debug`) has become (it is dropped/deprecated/still available).

File: docs/source/reference/core/configuration.rst [hunk 8]
```diff
@@ -0,0 +1,102 @@
...
+``chainer.config.debug``
+   Debug mode flag.
+   If it is ``True``, Chainer runs in the debug mode.
+   See :ref:`debug` for more information of the debug mode.
+   The default value is given by ``CHAINER_DEBUG`` environment variable (set to 0 or 1) if available, otherwise uses ``False``.
+``chainer.config.enable_backprop``
+   Flag to enable backpropagation support.
+   If it is ``True``, the default behavior of :class:`Function` application to :class:`Variable` is non-volatile if all inputs has ``AUTO`` volatile flag.
+   Otherwise, the default behavior is set to volatile mode.
```
Comment (mitmul) [inline docs/source/reference/core/configuration.rst]: How about adding an explanation about "volatile mode"? For example,

~is set to volatile mode that doesn't keep track of any function applications in all :class:`Variable`s.

File: docs/source/tutorial/basic.rst (hunk 6 above)
Comment (beam2d) [inline docs/source/tutorial/basic.rst]: Yes, I meant that. I will revise this part. Thank you.

File: docs/source/reference/core/configuration.rst (hunk 7 above)
Comment (beam2d) [inline docs/source/reference/core/configuration.rst]: I see, that makes sense. I will add it. Thank you.

File: docs/source/reference/core/configuration.rst (hunk 8 above)
Comment (beam2d) [inline docs/source/reference/core/configuration.rst]: I'll add.

**Comment by beam2d** [comment]
@delta2323 The documentation of the conventional ways of configuring debug mode is left in debug.rst.

**Comment by beam2d** [comment]
Thank you for your comments. I updated the document. I also improve the structure of the second paragraph of configuration.rst.

**Comment by delta2323** [comment]
@beam2d Sorry, I overlooked about the conventional way. That seems good. Thank you.

**Comment by delta2323** [comment]
LGTM

**Comment by delta2323** [comment]
LGTM

**Comment by mitmul** [comment]
LGTM

File: chainer/configuration.py [hunk 9]
```diff
@@ -69,36 +50,65 @@ def __setattr__(self, name, value):
...
+        Prints the config entries.
 
         The entries are sorted in the lexicographical order of the entry names.
 
         Args:
             file: Output file-like object.
 
+        .. admonition:: Example
+
+           You can easily print the list of configurations used in the current thread.
```
Comment (mitmul) [inline chainer/configuration.py]: This line seems to exceed 79 characters.

Comment (delta2323) [inline chainer/configuration.py]: I fixed it.

**Comment by bkvogel** [comment]
LGTM!

**Pull request merged**